language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,245 @@
+                    if removed_edge.edge_type == EdgeType::Strong {
+                        let mut new_label = vec![];
+                        new_label.append(&mut removed_edge.label.clone());
+                        new_label.append(&mut nonce_edge.label.clone());
+                        let edge = Edge {
+                            edge_type: nonce_edge.edge_type,
+                            label: new_label,
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    } else {
+                        let edge = Edge {
+                            edge_type: EdgeType::Weak,
+                            label: removed_edge.label.clone(),
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    }
+                });
+            });
+        }
+        checked_verify!(self.invariant());
+        checked_postcondition!(!self.0.contains_key(&nonce));
+    }
+
+    /// renames nonces in `self` according to `nonce_map`
+    pub fn rename_nonces(&self, nonce_map: BTreeMap<Nonce, Nonce>) -> Self {
+        checked_assume!(self.invariant());
+        let mut new_graph = BTreeMap::new();
+        for (n, es) in &self.0 {
+            new_graph.insert(
+                nonce_map[n],
+                es.iter()
+                    .map(|x| Edge {
+                        edge_type: x.edge_type,
+                        label: x.label.clone(),
+                        to: nonce_map[&x.to],
+                    })
+                    .collect::<BTreeSet<_>>(),
+            );
+        }
+        let new_borrow_graph = BorrowGraph(new_graph);
+        checked_verify!(self.invariant());
+        checked_verify!(new_borrow_graph.invariant());
+        new_borrow_graph
+    }
+
+    /// checks if `self` covers `other`
+    pub fn abstracts(&self, other: &Self) -> bool {
+        self.unmatched_edges(other).values().all(|es| es.is_empty())
+    }
+
+    /// joins `other` into `self`
+    pub fn join(&mut self, other: &Self) {
+        for (n, es) in self.unmatched_edges(other) {
+            es.into_iter().for_each(|e| {
+                self.0.get_mut(&n).unwrap().insert(e);
+            })
+        }
+    }
+
+    /// gets all nonces that are targets of outgoing edges from `nonce`
+    pub fn all_borrows(&self, nonce: Nonce) -> BTreeSet<Nonce> {
+        checked_precondition!(self.0.contains_key(&nonce));
+        self.0[&nonce].iter().map(|x| x.to).collect()
+    }
+
+    /// gets all nonces that are targets of outgoing edges from `nonce` that are labeled with the empty label
+    pub fn nil_borrows(&self, nonce: Nonce) -> BTreeSet<Nonce> {
+        checked_precondition!(self.0.contains_key(&nonce));
+        self.0[&nonce]
+            .iter()
+            .filter(|x| x.label.is_empty())
+            .map(|x| x.to)
```

Comment:
Is this comment about performance or code clarity? 

I find the current idiom clearer than the filter_map idiom because I have a general Boolean condition in the filter  which does not naturally connected to an Option.

Reply:
Originally when I saw this, I thought "Ah! Yes, filter_map could be nice "

But now I have to agree with @shazqadeer, the filter + a map is much easier to reason about. 

Performance wise, they should be exactly the same (if anything the filter + map should be faster as I imagine allocation will dominate here).
- I imagine the iterators get deforested and optimized so there should be no discernible performance difference
- This code does not need to be hyper optimized. So regardless, I would pick the more readable implementation, which in my opinion is the filter + map

Reply:
I was just suggesting *giving some thought* to 
```
.filter_map(|x| if x.label.is_empty() {Some(x.to)} else {None})
```
I'm more than happy to defer to you both on readability.